File: frontend_admin/src/services/roles.ts
import api from "@/lib/axios";

// 角色相关类型定义
export interface Role {
  id: number;
  name: string;
  description?: string;
  permissions?: string[];
}

export interface RoleCreate {
  name: string;
  description?: string;
  permissions?: string[];
}

export interface RoleUpdate {
  name?: string;
  description?: string;
  permissions?: string[];
}

/**
 * 角色管理 API 服务
 */
export const roleService = {
  /**
   * 获取角色列表
   * @param params 查询参数
   */
  getRoles: (params?: { skip?: number; limit?: number }) => {
    return api.get<Role[]>('/admin/roles/', { params });
  },

  /**
   * 创建角色
   * @param data 角色数据
   */
  createRole: (data: RoleCreate) => {
    return api.post<Role>('/admin/roles/', data);
  },

  /**
   * 更新角色
   * @param id 角色ID
   * @param data 更新数据
   */
  updateRole: (id: number, data: RoleUpdate) => {
    return api.put<Role>(`/admin/roles/${id}`, data);
  },

  /**
   * 删除角色
   * @param id 角色ID
   */
  deleteRole: (id: number) => {
    return api.delete<Role>(`/admin/roles/${id}`);
  },
};

File: backend/app/api/api_v1/endpoints/admin/roles.py
from typing import Any, List

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app import models, schemas, crud
from app.api import deps
from app.schemas.response import ResponseSuccess
from app.core.exceptions import CustomException

router = APIRouter()

@router.get("/", response_model=ResponseSuccess[List[schemas.Role]])
def read_roles(
    db: Session = Depends(deps.get_db),
    skip: int = 0,
    limit: int = 100,
    current_user: models.User = Depends(deps.get_current_active_superuser),
) -> Any:
    """
    获取角色列表
    Retrieve roles.
    """
    roles = crud.role.get_multi(db, skip=skip, limit=limit)
    return ResponseSuccess(data=roles)

@router.post("/", response_model=ResponseSuccess[schemas.Role])
def create_role(
    *,
    db: Session = Depends(deps.get_db),
    role_in: schemas.RoleCreate,
    current_user: models.User = Depends(deps.get_current_active_superuser),
) -> Any:
    """
    创建新角色
    Create new role.
    """
    role = crud.role.get_by_name(db, name=role_in.name)
    if role:
        raise CustomException(
            code=400,
            message="The role with this name already exists in the system.", # 该名称的角色已存在
        )
    role = crud.role.create(db, obj_in=role_in)
    return ResponseSuccess(data=role)

@router.put("/{role_id}", response_model=ResponseSuccess[schemas.Role])
def update_role(
    *,
    db: Session = Depends(deps.get_db),
    role_id: int,
    role_in: schemas.RoleUpdate,
    current_user: models.User = Depends(deps.get_current_active_superuser),
) -> Any:
    """
    更新角色
    Update a role.
    """
    role = crud.role.get(db, id=role_id)
    if not role:
        raise CustomException(
            code=404,
            message="The role with this id does not exist in the system", # 该ID的角色不存在
        )
    role = crud.role.update(db, db_obj=role, obj_in=role_in)
    return ResponseSuccess(data=role)

@router.delete("/{role_id}", response_model=ResponseSuccess[schemas.Role])
def delete_role(
    *,
    db: Session = Depends(deps.get_db),
    role_id: int,
    current_user: models.User = Depends(deps.get_current_active_superuser),
) -> Any:
    """
    删除角色
    Delete a role.
    """
    role = crud.role.get(db, id=role_id)
    if not role:
        raise CustomException(
            code=404,
            message="The role with this id does not exist in the system", # 该ID的角色不存在
        )
    role = crud.role.remove(db, id=role_id)
    return ResponseSuccess(data=role)
